scipy/special/special/legendre.h

```diff
@@ -1,352 +1,998 @@
+    init_n(res);
+
+    legendre_p_recurrence_n<T> re_n{z};
+    forward_recur(0, n + 1, re_n, res, f);
 }
 
-template <typename T, typename InputVec, typename OutputVec>
-void legendre_all_jac(T z, InputVec p, OutputVec pd) {
-    long n = p.extent(0) - 1;
+/**
+ * Compute the Legendre polynomial of degree n.
+ *
+ * @param n degree of the polynomial
+ * @param z argument of the polynomial, either real or complex
+ * @param res result (gradient)
+ */
+template <typename T, typename... OutputVals>
+void legendre_p(int n, T z, std::tuple<OutputVals &...> res) {
+    static constexpr size_t N = sizeof...(OutputVals) - 1;
+
+    grad_tuple_t<T[2], N> res_n;
+    legendre_p_for_each_n(n, z, tuples::ref(res_n), [](int n, grad_tuple_t<T(&)[2], N> res_n) {});
+
+    res = tuples::access(res_n, 1);
+}
 
-    pd(0) = 0;
-    if (n >= 1) {
-        pd(1) = 1;
+/**
+ * Compute the Legendre polynomial of degree n.
+ *
+ * @param n degree of the polynomial
+ * @param z argument of the polynomial, either real or complex
+ *
+ * @return value of the polynomial
+ */
+template <typename T>
+T legendre_p(int n, T z) {
+    T res;
+    legendre_p(n, z, std::tie(res));
+
+    return res;
+}
 
-        if (std::abs(std::real(z)) == 1 && std::imag(z) == 0) {
-            for (long k = 2; k <= n; k++) {
-                pd(k) = k * (k + 1) * std::pow(std::real(z), k + 1) / 2;
+/**
+ * Compute all Legendre polynomials of degree j, where 0 <= j <= n.
+ *
+ * @param z argument of the polynomials, either real or complex
+ * @param res a view into a multidimensional array with element type T and size n + 1 to store the value of each
+ *            polynomial
+ */
+template <typename T, typename... OutputVecs>
+void legendre_p_all(T z, std::tuple<OutputVecs &...> res) {
+    static constexpr size_t N = sizeof...(OutputVecs) - 1;
+
+    auto &res0 = std::get<0>(res);
+    int n = res0.extent(0) - 1;
+
+    grad_tuple_t<T[2], N> res_n;
+    legendre_p_for_each_n(n, z, tuples::ref(res_n), [&res](int n, grad_tuple_t<T(&)[2], N> res_n) {
+        tuples::call(res, n) = tuples::access(res_n, 1);
+    });
+}
+
+struct assoc_legendre_unnorm_policy {};
+
+struct assoc_legendre_norm_policy {};
+
+constexpr assoc_legendre_unnorm_policy assoc_legendre_unnorm;
+
+constexpr assoc_legendre_norm_policy assoc_legendre_norm;
+
+template <typename T, typename NormPolicy>
+struct assoc_legendre_p_initializer_m_abs_m;
+
+template <typename T>
+struct assoc_legendre_p_initializer_m_abs_m<T, assoc_legendre_unnorm_policy> {
+    bool m_signbit;
+    T z;
+    int branch_cut;
+    T branch_cut_sign;
+    T w;
+
+    assoc_legendre_p_initializer_m_abs_m(bool m_signbit, T z, int branch_cut)
+        : m_signbit(m_signbit), z(z), branch_cut(branch_cut) {
+        if (branch_cut == 3) {
+            branch_cut_sign = -1;
+
+            w = std::sqrt(z * z - T(1)); // do not modify, see function comment
+            if (std::real(z) < 0) {
+                w = -w;
             }
         } else {
```

Comment:
When `branch_cut` isn't one of 2 or 3, it's just automatically treated as 2 here. I think it would be better for `assoc_legendre_p` and `assoc_legendre_p_all` to return `nan` in those cases and use `set_error` to give an informative message.

Reply:
I plan to do this in a follow up. The current broken behavior is lifted from `clpmn`.